java/src/main/native/ai_onnxruntime_OrtTrainingSession.c

```diff
@@ -0,0 +1,759 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates. All rights reserved.
+ * Licensed under the MIT License.
+ */
+#include <jni.h>
+#include <string.h>
+#include <stdlib.h>
+#include "OrtJniUtil.h"
+#include "onnxruntime/core/session/onnxruntime_c_api.h"
+#include "onnxruntime_training_c_api.h"
+#include "ai_onnxruntime_OrtTrainingSession.h"
+
+#ifdef _WIN32
+wchar_t* copyAndPad(JNIEnv * jniEnv, jstring javaStr) {
+  // The output of GetStringChars is not null-terminated, so we copy it and add a terminator
+  const jchar* charArr = (*jniEnv)->GetStringChars(jniEnv, javaStr, NULL);
+  size_t strLength = (*jniEnv)->GetStringLength(jniEnv, javaStr);
+  wchar_t* outputStr = (wchar_t*)calloc(strLength + 1, sizeof(wchar_t));
+  if (outputStr != NULL) {
+    wcsncpy_s(outputStr, outputStrLength + 1, (const wchar_t*)charArr, strLength);
```

Comment:
> wcsncpy_s(outputStr, outputStrLength + 1, (const wchar_t*)charArr, strLength); [](http://example.com/codeflow?start=3&length=79)

The source does not have null terminator.
I have studied the docs and the function will not add a null terminator except by accident, that is the extra + 1 character contains null.
So it has to be added manually. Like:
```.c
outputStr[outputStrLength] = 0;
```

Reply:
That's why the memory is allocated with calloc to init everything to zero. I can move it to malloc and explicitly set the zero byte if you want.

Reply:
No need. I am forgetting C.